RPC: Export GUI Policy Options for Headless Deployments

kwsantiago opened the pull request:
## Summary

Add RPC commands to export/import GUI policy settings for headless systems (Start9, Umbrel).
Relates to https://github.com/bitcoinknots/bitcoin/issues/130

## Changes

- `dumpsettings`: Export settings to JSON
- `getsettings`: Query specific settings  
- `getsettingsschema`: Generate JSON Forms schema
- `setsetting`: Update individual settings
- `updatesettings`: Bulk atomic updates
- `subscribesettings`: Change notifications

GUI export/import buttons and ZMQ notifications included.

## Test plan

- RPC commands work with bitcoin-cli
- GUI export/import functionality
- Functional and unit tests pass

luke-jr commented on doc/release-notes.md:
```diff
@@ -49,6 +49,37 @@ to do so until/unless that is resolved.
 Notable changes
 ===============
 
+Settings Export System
```
Please put this in a file named `doc/release-notes-k154.md`

luke-jr commented on doc/settings-export.md:
```diff
@@ -0,0 +1,87 @@
+
+**Arguments:**
+- `category` (string, optional): Filter by category ("wallet", "mempool", "relay", "script", "transaction", "data_carrier", "dust", "block_creation", "network", "gui")
+- `include_sensitive` (boolean, optional, default=false): Include sensitive settings
```
Maybe it shouldn't be allowed at all for security?

luke-jr commented on doc/settings-export.md:
```diff
@@ -0,0 +1,87 @@
+
+Export settings to JSON format.
+
+**Syntax:** `dumpsettings ( "category" include_sensitive "encrypt_password" )`
```
At least the boolean should probably be named-only

luke-jr commented on doc/settings-export.md:
```diff
@@ -0,0 +1,87 @@
+**Arguments:**
+- `category` (string, optional): Filter by category ("wallet", "mempool", "relay", "script", "transaction", "data_carrier", "dust", "block_creation", "network", "gui")
+- `include_sensitive` (boolean, optional, default=false): Include sensitive settings
+- `encrypt_password` (string, optional): Password to encrypt output
```
This feels a bit overengineered. What is the use case for an untrusted JSON-RPC connection? If an attacker can MITM it, they can just change the password in the request...

luke-jr commented on doc/settings-export.md:
````diff
@@ -0,0 +1,87 @@
+bitcoin-cli dumpsettings "" true "password"
+```
+
+### getsettings
````
Feels a bit redundant with dumpsettings. Probably should just have one of these (with parameters controlling what they return)?

luke-jr commented on doc/settings-export.md:
````diff
@@ -0,0 +1,87 @@
+bitcoin-cli getsettingsschema "wallet"
+```
+
+### setsetting
````
Probably should just have `setsettings`

luke-jr commented on doc/settings-security.md:
```diff
@@ -0,0 +1,70 @@
+
+### Write Permissions
+- `settings-write`: Modify non-critical settings
+- `settings-write-critical`: Modify critical settings (network, security)
```
Maybe setting groups can be reused for this?

luke-jr commented on doc/settings-security.md:
```diff
@@ -0,0 +1,70 @@
+
+## Rate Limiting
+
+- 50 setting changes per 5 minutes per user
```
Concept NACK, let the users do what they want.

luke-jr commented on src/qt/optionsdialog.cpp:
```diff
@@ -1048,6 +1055,146 @@ void OptionsDialog::on_cancelButton_clicked()
+        if (!previewResult.changes.isEmpty()) {
+            QString changesList;
+            for (const auto& change : previewResult.changes) {
+                changesList += QString("• %1: %2 → %3\n")
```
Probably want something more user-friendly here, potentially listing dangerous changes with an icon at the top?

luke-jr commented on src/qt/optionsdialog.cpp:
```diff
@@ -1048,6 +1055,146 @@ void OptionsDialog::on_cancelButton_clicked()
+                tr("Import Settings Preview"),
+                tr("The following settings will be changed:\n\n%1\n"
+                   "Do you want to proceed with the import?").arg(changesList),
+                QMessageBox::Yes | QMessageBox::No
```
Might make sense to use the countdown timer here. Maybe even two of them if dangerous settings are being touched.

luke-jr reviewed (commented):
There's a lot to review here, so this isn't complete comments yet.

kwsantiago commented:
Thank you for taking a look @luke-jr I greatly appreciate it.

This is still a WIP and there's more I'd like to do here, but your comments are really helpful as I work through this to ensure it meets Knots' standards.

luke-jr commented:
One issue is going to be review. This is largely one giant commit, which makes it difficult. Can you break it up into smaller steps? 

You're also putting a lot in the RPC code. At the end of this, we should ideally have RPC and GUI using the same code for changing settings. So an internal interface for updatesettings called by both would be appropriate. There's a lot of existing parsing code that you can probably reuse too (eg, boolean string to boolean UniValue conversions...)

kwsantiago commented:
@luke-jr for review purposes, would it be better to break this into several PRs, or would you rather have everything in one PR?

luke-jr commented:
Multiple PRs could make sense, but there's tradeoffs. As long as it's broken up into logical separate commits, that's probably good enough.

luke-jr commented on src/common/settings_json.cpp:
```diff
@@ -0,0 +1,468 @@
+
+namespace common {
+
+// Define type information for all settings from OptionsModel
```
This is largely redundant with all the `AddArg` calls - can we reuse those?

pdath commented on src/node/settings_manager.cpp:
```diff
@@ -0,0 +1,419 @@
+
+void SettingsManager::RegisterRuntimeHandler(const std::string& setting_name,
+                                            std::function<bool(const UniValue&)> handler) {
+    m_impl->m_runtime_handlers[setting_name] = handler;
```
From Code Rabbit:

m_runtime_handlers map accessed without holding m_settings_mutex.

Every Apply* handler takes LOCK(m_settings_mutex) before mutating shared state, but CanApplyAtRuntime, the find() in ApplySetting, and RegisterRuntimeHandler's map write all touch m_runtime_handlers unguarded. If RegisterRuntimeHandler is ever called concurrently with an in-flight ApplySetting (e.g. RPC + wallet/init wiring), this is a data race on std::map.

pdath commented on src/node/settings_manager.cpp:
```diff
@@ -0,0 +1,419 @@
+
+        m_runtime_handlers["rejectbaremultisig"] = [this](const UniValue& value) {
+            return ApplyRejectPolicy("rejectbaremultisig", value.get_bool());
+        };
```
From Code Rabbit:

reject* **runtime settings don't drive the active policy** ApplyRejectPolicy() only writes -reject... entries and flips g_reject_*, but relay/mining still reads -permitbarepubkey, -permitbaremultisig, and -spkreuse/SpkReuseMode; the new atomics are not used by the policy paths. Update the live mempool/policy state here, not just the shadow settings.

pdath commented on src/node/settings_manager.cpp:
```diff
@@ -0,0 +1,419 @@
+            return true;
+        }
+        return false;
+    }
```
From Code Rabbit:

Runtime setters need to update the live policy state

* ApplyBytesPerSigOp, ApplyDataCarrierCost, ApplyMinRelayTxFee, ApplyIncrementalRelayFee, and ApplyDustRelayFee only persist the arg and emit a notification; the live relay/mining paths read nBytesPerSigOp, g_weight_per_data_byte, and m_node.mempool->m_opts.*, so these changes won’t affect current behavior until restart.
* ApplyDataCarrierCost also uses std::stod; switch to locale-independent fixed-point parsing instead.

pdath commented on src/node/settings_manager.cpp:
```diff
@@ -0,0 +1,419 @@
+
+        file.close();
+        return true;
+    }
```
From Code Rabbit:

Persist through the writable settings path

* PersistToFile is mutating gArgs.GetConfigFilePath(), but -conf is documented as read-only. Use the existing writable settings/config helpers (-settings / -confrw) instead.
* This read-modify-write is also unguarded, so concurrent RPC updates can overwrite each other.

pdath commented on src/node/settings_manager.cpp:
```diff
@@ -0,0 +1,419 @@
+            return true;
+        }
+        return false;
+    }
```
From Code Rabbit:
Widen the mempool limit update and take the mempool lock.

* m_opts.max_size_bytes/m_opts.expiry are read under mempool.cs in validation paths; mutating them here without that lock can race with trimming/expiry.
* size_mb * 1'000'000 should be widened to int64_t first; valid values above ~2147 MB overflow before assignment.

pdath commented on src/common/settings_json.cpp:
```diff
@@ -0,0 +1,205 @@
+    return {type, 0, 0, {},
+            metadata.exists("description") ? metadata["description"].get_str() : "Unknown setting",
+            metadata.exists("restart_required") ? metadata["restart_required"].get_bool() : false,
+            SettingCategory::NODE};
```
From Code Rabbit:
GetSettingTypeInfo **drops constraint data and mis-reports category**.

min_value/max_value/allowed_values are always left at their defaults (0, 0, {}) even though metadata["constraints"] (populated via GetSettingMetadata) can carry min/max/enum. category is unconditionally SettingCategory::NODE, ignoring metadata["category"]. Any consumer relying on this struct for validation UI/bounds will silently get wrong/empty bounds and wrong category for every setting.

pdath commented on src/common/settings_json.cpp:
```diff
@@ -0,0 +1,205 @@
+        categories.pushKV(key, arr);
+    }
+
+    return categories;
```
From Code Rabbit:
GetSettingCategories() **is a hand-maintained, incomplete duplicate of** common::GetAllSettingsMetadata().

This hardcodes only a subset of settings/categories and is missing entire categories present in settings_metadata.cpp ("mining", "prune") as well as many individual settings within the categories it does list (e.g. mempoolexpiry, incrementalrelayfee, datacarriersize, blockmaxweight, port, prune, etc.). Consumers of this function will get an incomplete settings catalog. Consider deriving it from common::GetAllSettingsMetadata() (grouping by metadata.category) so it can't drift.

pdath commented on src/common/settings_json.cpp:
```diff
@@ -0,0 +1,205 @@
+
+bool JsonToSettings(const UniValue& json, Settings& settings, std::vector<std::string>& errors)
+{
+    return true;
```
From Code Rabbit:
SettingsToJson/JsonToSettings **are non-functional stubs**.

SettingsToJson only emits {"version": 1} — it never serializes any actual setting from settings. JsonToSettings always returns true without reading json or populating settings/errors. Per the stack context, the downstream RPC layer (dumpsettings/getsettings/setsettings) depends on this contract for export/import — as written, export/import of settings data is a no-op.

pdath commented on src/qt/optionsmodel.cpp:
```diff
@@ -1614,3 +1636,293 @@ void OptionsModel::checkAndMigrate()
+                    optionIndex = i;
+                    break;
+                }
+            }
```
From Code Rabbit:
**Same** SettingName() **throw breaks** previewSettingsImport().

This inner lookup calls SettingName(static_cast<OptionID>(i)) for i starting at 0, which throws std::logic_error for GUI‑only options before any match is found. The outer try/catch then marks the preview invalid, so preview (and therefore importSettings(), which calls it first) always fails whenever the settings object is non‑empty. Wrap the call in a try/catch that continues on std::logic_error, mirroring the export fix.

pdath commented on src/qt/optionsmodel.cpp:
```diff
@@ -1614,3 +1636,293 @@ void OptionsModel::checkAndMigrate()
+            // Find corresponding option
+            for (int i = 0; i < OptionIDRowCount; ++i) {
+                OptionID option = static_cast<OptionID>(i);
+                if (SettingName(option) == settingName) {
```
From Code Rabbit:
**Same** SettingName() **throw breaks** importSettings().

Identical root cause: SettingName(option) throws for GUI‑only options in this lookup loop, and the outer try/catch turns it into an import error. Guard the call with try/catch on std::logic_error (or reuse a shared name→OptionID lookup helper) so unmapped options are skipped rather than aborting the whole import.

pdath commented on src/qt/optionsmodel.cpp:
```diff
@@ -1614,3 +1636,293 @@ void OptionsModel::checkAndMigrate()
+                    settings.pushKV(settingName.toStdString(), value.toString().toStdString());
+                    break;
+            }
+        }
```
From Code Rabbit:
**Critical**: exportSettings() **always fails** — SettingName() **throws for GUI‑only options**.

SettingName() (Lines 59‑124) throws std::logic_error in its default: branch for every option that has no corresponding node setting (e.g. StartAtStartup, which is index 0, plus ShowTrayIcon, MinimizeToTray, NetworkPort, DisplayUnit, corepolicy, etc.). This loop iterates 0 .. OptionIDRowCount, so it throws on the very first iteration; the surrounding try/catch converts it into a std::runtime_error and the whole export aborts. Export can never succeed.

Skip GUI‑only options instead of letting SettingName() throw the entire operation.

pdath commented on src/qt/optionsmodel.cpp:
```diff
@@ -306,6 +307,27 @@ bool OptionsModel::Init(bilingual_str& error)
+        // Handle the setting change on the Qt thread
+        QMetaObject::invokeMethod(this, "handleExternalSettingChange", Qt::QueuedConnection,
+                                 Q_ARG(QString, qSettingName), Q_ARG(QVariant, qNewValue));
+    });
```
From Code Rabbit:
**Keep the** NotifySettingChanged **connection alive** src/qt/optionsmodel.cpp:311
uiInterface.NotifySettingChanged_connect(...) returns a boost::signals2::connection, so dropping it leaves the callback registered indefinitely. If OptionsModel is destroyed first, a later setting change can call back into a dangling this; store the handle as a member and disconnect it during teardown.

pdath commented on src/rpc/settings.cpp:
```diff
@@ -0,0 +1,1116 @@
+    result.pushKV("poll_token", poll_token);
+
+    bool has_changes = since_token.empty() || (since_token != poll_token);
+    result.pushKV("has_changes", has_changes);
```
From Code Rabbit:
**Poll token is time-derived, so** has_changes **reports false positives**.

hasher << current_time / 10 mixes wall-clock time into the token. The token therefore rotates roughly every 10 seconds regardless of whether any setting actually changed, so a client polling with a since_token older than that window will always see has_changes = true. This defeats the documented change-detection purpose ("compare the returned token to detect setting changes"). Hash only the setting values (and category filter), not the timestamp.

pdath commented on src/common/settings_metadata.cpp:
```diff
@@ -0,0 +1,838 @@
+        }(),
+        false,
+        false
+    };
```
From Code Rabbit:
mintxfee **min constraint must be a string** (src/common/settings_metadata.cpp:46-61) — ValidateSettingValue() reads amount constraints with get_str(), so c.pushKV("min", 0) will throw for this setting. Change it to "0" (or make the amount-constraint reads use getValStr()).

pdath commented on src/rpc/settings_schema.cpp:
```diff
@@ -0,0 +1,671 @@
+    mempool_data.pushKV("mempooltruc", true);
+    form_data.pushKV("mempool", mempool_data);
+
+    result.pushKV("formData", form_data);
```
From Code Rabbit:
formData **returns hardcoded values instead of the node's actual settings**.

GenerateJSONFormsSchema accepts node_context and args but never reads them; form_data is populated with fabricated constants (e.g. walletrbf=true, maxmempool=300) as the in-code comment admits. A settings UI consuming getsettingsschema will render wrong current values. Populate formData from args/settings, or drop the unused parameters and document that formData contains defaults only.

pdath commented on test/functional/rpc_settings_security.py:
```diff
@@ -0,0 +1,84 @@
+
+        for setting in critical_settings:
+            self.log.info(f"Testing critical setting: {setting}")
+
```
From Code Rabbit:
**This helper never verifies protection behavior.**

test_critical_settings only logs names and never performs an RPC or asserts authorization, so it cannot catch regressions in critical-setting protection.

pdath commented on test/functional/rpc_settings_security.py:
```diff
@@ -0,0 +1,84 @@
+
+        except Exception as e:
+            self.log.error(f"Error testing audit logging: {e}")
+            raise
```
From Code Rabbit:
**Don't swallow RPC failures here.**

These blanket except Exception blocks let the workflow keep going when setsetting, setsettings, or the notification check breaks, so the test can still pass without exercising the new RPC paths.

Also applies to: 97-106

pdath commented on test/functional/rpc_settings.py:
```diff
@@ -0,0 +1,980 @@
+                else:
+                    self.log.info(f"{test_name}: Failed with: {str(e)}")
+
+        self.log.info("Error handling robustness test completed")
```
From Code Rabbit:
test_error_handling_robustness **cannot ever fail**.

Every branch here (RPC call succeeds, raises an "error"/"invalid"-labeled exception, or raises any other exception) just logs an info message — none of the three branches calls a failing assertion. This test provides zero real coverage of error-handling correctness; it will pass unconditionally regardless of actual RPC behavior.

pdath commented on test/functional/rpc_settings.py:
```diff
@@ -0,0 +1,980 @@
+                assert_raises_rpc_error(-8, "Invalid category", node.dumpsettings, special_cat)
+                self.log.info(f"Special character category '{special_cat}' correctly rejected")
+            except Exception as e:
+                self.log.info(f"Special category '{special_cat}' test: {str(e)}")
```
From Code Rabbit:
**Remove the broad** except Exception **around these RPC assertions** — assert_raises_rpc_error() raises AssertionError when the expected RPC error is missing or mismatched, so these handlers turn a failing check into an info log. This applies here and at lines 634-646; drop the wrapper or narrow it to the RPC call only.

pdath commented on test/functional/rpc_settings.py:
```diff
@@ -0,0 +1,980 @@
+                        self.log.info("No changes detected in polling (expected if no actual change occurred)")
+
+        except Exception as e:
+            self.log.info(f"Settings change test skipped: {str(e)}")
```
**From Code Rabbit:
Systemic broad exception-swallowing weakens these tests' regression-catching value.**

Nearly every test method in this file wraps its core assertions in try/except Exception as e: self.log.info(...), converting what should be hard test failures into soft log messages when the settings RPCs misbehave (e.g. test_settings_notifications, test_rpc_permissions, test_concurrent_access, test_stress_testing, test_boundary_conditions, test_json_injection_safety, test_data_integrity). This means regressions in setsetting/setsettings/subscribesettings behavior may go undetected by CI even though this test "passes."

Consider narrowing exception handling to specific expected exception types and using assert_raises_rpc_error/hard assertions for negative-path checks, reserving try/except for genuinely optional/environment-dependent RPCs only.

Also applies to: 167-205, 206-257, 259-297, 299-344, 345-389, 390-443**

pdath commented on src/interfaces/settings_notifications.cpp:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        LogDebug(BCLog::RPC, "Settings notification: %s changed from %s to %s (source: %s)\n",
+                setting_name, old_value.write(), new_value.write(), source);
```
From Code Rabbit:
Sensitive setting values are logged unmasked.

SettingMetadata::is_sensitive (settings_metadata.h) exists specifically to mark values that "should be masked in exports," but NotifySettingChanged logs old_value/new_value unconditionally regardless of this flag. Consider checking the setting's metadata and redacting the value when is_sensitive is true.

pdath commented on src/interfaces/settings_notifications.cpp:
```diff
@@ -0,0 +1,156 @@
+                );
+            }
+        );
+    }
```
From Code Rabbit:
**Redact sensitive values before logging setting changes** src/interfaces/settings_notifications.cpp:93-94

old_value.write() and new_value.write() log raw values for every update, including sensitive settings like rpcpassword and walletpassphrase. Mask them here before writing to the RPC debug log.

pdath commented on test/functional/rpc_settings_security.py:
```diff
@@ -0,0 +1,84 @@
+        assert "settings" in result
+
+        result_detailed = node.dumpsettings("", {"detailed": True})
+        assert "settings" in result_detailed
```
From Code Rabbit:
**Assert masking behavior, not just response shape.**

These checks only verify that dumpsettings returns a settings key, so a regression that leaks sensitive values would still pass. Validate that known sensitive fields stay redacted by default and only appear when explicitly requested.

pdath commented on test/functional/rpc_settings_security.py:
```diff
@@ -0,0 +1,84 @@
+        critical_settings = ["rpcport", "bind", "port", "maxconnections"]
+
+        for setting in critical_settings:
+            self.log.info(f"Testing critical setting: {setting}")
```
From Code Rabbit:
**This test is currently a no-op.**

The loop only logs the candidate settings and never exercises setsetting or any permission check, so it cannot fail even if critical-setting protection breaks.

pdath commented on doc/settings-security.md:
````diff
@@ -0,0 +1,73 @@
+# Full access (including critical network/rpc settings)
+bitcoind -rpcwhitelist=admin:dumpsettings,getsettingsschema,subscribesettings,setsettings,setsettings \
+         -rpcwhitelistpermissions=admin:settings-write:network,settings-write:rpc
+```
````
**Fix the whitelist examples.**

Both examples duplicate setsettings, and neither includes the base settings-write permission even though line 12 says it is required for all modifications. As written, the samples contradict the permission model above.

pdath commented on src/common/settings_json.h:
```diff
@@ -0,0 +1,140 @@
+    NETWORK,
+    GUI,
+    NODE
+};
```
From Code Rabbit:
**SettingCategory enum is missing categories used elsewhere.**

settings_metadata.cpp defines settings with category "mining" (corepolicy) and "prune" (prune, txindex, blockfilterindex, coinstatsindex), but this enum has no MINING/PRUNE value. Any code mapping metadata category strings to this enum will mis-categorize (or fail to categorize) those settings.

pdath commented on src/init.cpp:
```diff
@@ -1947,6 +1947,13 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                UniValue old_value; // TODO: Pass actual old value from caller
+                g_zmq_notification_interface->SettingChanged(setting_name, old_value, new_value, "RPC");
+            }
+        });
```
**From Code Rabbit:
ZMQ setting-changed forwarding always reports an empty old value and a hardcoded "RPC" source.**

UniValue old_value; is default-constructed (null) every time and never populated with the actual previous value, and the source is unconditionally "RPC" regardless of where the change actually originated (GUI, config reload, etc.). The // TODO comment confirms this is a known gap — any ZMQ consumer relying on the old-value diff or accurate origin will get wrong data.

Perhaps threading the real previous value and origin through NotifySettingChanged/SettingsNotifications so this isn't hardcoded?

pdath commented on src/interfaces/settings_notifications.h:
```diff
@@ -0,0 +1,71 @@
+                          const UniValue& old_value,
+                          const UniValue& new_value,
+                          const std::string& source)> callback,
+        const std::string& category = "") = 0;
```
From Code Rabbit:
**Mark** Subscribe **as** [[nodiscard]].

The returned handle is the subscription (its destructor unsubscribes). If a caller writes notifications->Subscribe(cb); without capturing the result, the temporary is destroyed at the end of the full expression and the subscription is immediately cancelled — silently and without a compiler warning.

pdath commented on src/test/settings_notifications_tests.cpp:
```diff
@@ -0,0 +1,315 @@
+
+    std::cout << "Performance test completed in " << duration.count()
+              << "ms for " << num_subscribers << " subscribers and "
+              << num_notifications << " notifications" << std::endl;
```
From Code Rabbit:
**Unbounded busy-wait can hang the test suite indefinitely.**

while (total_callbacks.load() < num_subscribers * num_notifications) {
    std::this_thread::sleep_for(std::chrono::milliseconds(1));
}
If any notification is lost/dropped due to a bug in the implementation, this loop never terminates, hanging the entire test binary (and CI) rather than failing cleanly.

pdath commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -338,3 +341,20 @@ bool CZMQPublishRawWalletTransactionNotifier::NotifyWalletTransaction(const CTra
+
+    std::string json_str = notification.write();
+    return SendZmqMessage(MSG_SETTINGS, json_str.data(), json_str.size());
+}
```
From Code Rabbit:
**Double-encoded / type-losing JSON for** old_value/new_value.

old_value and new_value arrive here already JSON-serialized (via UniValue::write() in CZMQNotificationInterface::SettingChanged). Pushing them with pushKV(key, std::string) re-wraps that JSON text as a plain JSON string, so:

* String settings get double-quoted/escaped ("old_value":"\"foo\"").
* Numeric/bool/object settings lose their type entirely and become strings ("old_value":"5" instead of 5).
This diverges from the typed settings JSON contract used elsewhere in this stack and will confuse pubsettings ZMQ consumers.

kwsantiago commented:
Reworked onto current 29.x-knots: kept dumpsettings/setsettings only, routed both the RPC and the GUI through one shared internal interface (node/policy_settings) that applies to live mempool options and persists like the GUI, and added a functional test. Dropped the standalone settings manager, JSON schema/subscribe surface, and security levels.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,413 @@
+    return "accept";
+}
+
+std::string PermitEphemeralToString(const kernel::MemPoolOptions& o)
+{
+    if (!(o.permitephemeral_anchor || o.permitephemeral_send)) return "reject";
+    return std::string(o.permitephemeral_anchor ? "" : "-") + "anchor," +
+           (o.permitephemeral_send ? "" : "-") + "send," +
+           (o.permitephemeral_dust ? "" : "-") + "dust";
+}
```
From CodeRabbit, minor:

PermitEphemeralToString **drops** dust **for** -anchor,-send,dust setsettings accepts that combination via ApplyPermitEphemeralOption, but dumpsettings collapses it to reject, so dust won’t round-trip.

kwsantiago replied:
Fixed in 3baa9625: dust now round-trips (only collapses to `reject` when anchor, send and dust are all off).

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,413 @@
+            args.ModifyRWConfigFile("mempoolreplacement", nv.s);
+        }
+        return {};
+    }
```
From CodeRabbit, minor:

**Use the startup parser** for mempoolreplacement
This branch maps never to Never, any string containing optin (but not -optin) to OptIn, and everything else to Always. That diverges from the token-based startup parser, especially for values like fee and other comma lists, so setsettings can resolve the same string differently than CLI/config. Reuse the shared parser/helper here.

kwsantiago replied:
Fixed in 3baa9625: now parses via the same comma/plus token logic as ApplyArgsManOptions.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,413 @@
+    }
+
+    return util::Error{Untranslated(strprintf("unhandled setting: %s", name))};
+}
```
From CodeRabbit, critical:

**Protect live mempool policy updates**
UpdatePolicySetting mutates pool->m_opts at runtime, but readers like MaybeSendFeefilter() use m_mempool.GetMinFee() and m_mempool.m_opts.min_relay_feerate without m_mempool.cs. That makes setsettings a data race unless these writes are taken under the same lock or moved to atomics.

kwsantiago replied:
Pre-existing lock-free pattern shared with the GUI: OptionsModel::setOption writes m_opts.min_relay_feerate lock-free (optionsmodel.cpp:1277). Readers don't share one lock (validation under cs_main, MaybeSendFeefilter under g_msgproc_mutex), so a proper fix means making these fields atomic across validation/net_processing/GUI, which is broader than this PR.

pdath commented:
tACK b77bdc14c92b68bb1448a09000dc6bd4b3e88995

I reviewed the code with Code Rabbit, and added the three issues (1 critical, 2 minor) as comments.

I tested using the following commands on Ubuntu 24.04 running on ARMv8-A (64-bit):

git clone https://github.com/bitcoinknots/bitcoin.git pr154
cd pr154
git fetch origin pull/154/head:pr154
git switch pr154

```
cmake -B build \
-DRDTS_CONSENT=IMPLICIT \
-DBUILD_GUI=OFF \
-DCMAKE_BUILD_TYPE=Release
nice cmake --build build -j4

cd build
python3 test/functional/test_runner.py rpc_settings.py
```

The test passes.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,419 @@
+        }
+    });
+    args.WriteSettingsFile();
+}
```
From CodePilot:
Major: Propagate settings-file write failures.

WriteSettingsFile() returns a status, but Line 54 discards it. Callers consequently return success after only changing in-memory settings when persistence fails, despite persist=true.

Return a util::Result from the persistence helpers and propagate failures, including whether the live update was already applied.

Also applies to: 287-289

kwsantiago replied:
Fixed in 1781b8fa: PersistJson returns the write status and UpdatePolicySetting propagates it. ModifyRWConfigFile is void, so only the settings.json path can report.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,419 @@
+    if (n == "acceptnonstddatacarrier") return UniValue(o.accept_non_std_datacarrier);
+    if (n == "maxtxlegacysigops") return UniValue((int64_t)o.maxtxlegacysigops);
+    if (n == "dustrelayfee") return UniValue(UniValue::VNUM, FormatMoney(o.dust_relay_feerate_floor.GetFeePerK()));
+    // dustdynamic is read from the persisted setting string (matches the GUI); we
+    // persist it on every change, so this reflects the current value.
+    if (n == "dustdynamic") return UniValue(args.GetArg("-dustdynamic", DEFAULT_DUST_DYNAMIC));
```
From Code Rabbit:
Major: **Keep nonpersistent** dustdynamic **updates visible to reads**.

With persist=false, Lines 410-411 update the live options, but ReadPolicySetting still reads the old value from ArgsManager. A successful update is therefore immediately reported as unchanged.

Update the process-local argument value regardless of persistence, or derive the returned value from the live options.

```
     if (n == "dustdynamic") {
         auto parsed = ParseDustDynamicOpt(nv.s, /*max_fee_estimate_blocks=*/1008);
         if (!parsed) return util::Error{util::ErrorString(parsed)};
         o.dust_relay_target = parsed->first;
         o.dust_relay_multiplier = parsed->second;
+        args.ForceSetArg("-dustdynamic", nv.s);
         json("dustdynamic", UniValue(nv.s));
         return {};
     }
```
Also applies to: 407-413

kwsantiago replied:
Fixed in 1781b8fa: the persist parameter had no callers, so it's removed. Updates always persist, so reads can't go stale.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,419 @@
+    if (n == "bytespersigopstrict") { nBytesPerSigOpStrict = (unsigned int)nv.i; rwconf("bytespersigopstrict", strprintf("%d", nv.i)); return {}; }
+    if (n == "maxscriptsize") { g_script_size_policy_limit = (unsigned int)nv.i; json("maxscriptsize", UniValue(nv.i)); return {}; }
+    if (n == "spkreuse") { rwconf("spkreuse", nv.s); return {}; } // restart-required; no live field
+
```
From Code Rabbit:
Major: **Reject nonpersistent** spkreuse **updates**.

Because this setting is restart-only, persist=false makes Line 300 return success without changing either live or persisted state. Return an error for that combination instead.

```
-    if (n == "spkreuse") { rwconf("spkreuse", nv.s); return {}; }
+    if (n == "spkreuse") {
+        if (!persist) {
+            return util::Error{Untranslated("spkreuse requires persistence because it only takes effect after restart")};
+        }
+        rwconf("spkreuse", nv.s);
+        return {};
+    }
```

kwsantiago replied:
Fixed in 1781b8fa: persist parameter removed (no callers), so spkreuse always writes.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,419 @@
+    ArgsManager& args = *Assert(node.args);
+    CTxMemPool* pool = MemPool(node);
+    const std::string n{name};
+
+    // Settings not backed by live mempool options.
+    if (n == "datacarriercost") return UniValue(double(g_weight_per_data_byte) / WITNESS_SCALE_FACTOR);
+    if (n == "bytespersigop") return UniValue((int64_t)nBytesPerSigOp);
+    if (n == "bytespersigopstrict") return UniValue((int64_t)nBytesPerSigOpStrict);
+    if (n == "maxscriptsize") return UniValue((int64_t)g_script_size_policy_limit);
+    if (n == "spkreuse") return UniValue(args.GetArg("-spkreuse", "allow"));
+
+    if (!pool) return util::Error{Untranslated("mempool not available")};
+    const kernel::MemPoolOptions& o = pool->m_opts;
+
+    if (n == "maxmempool") return UniValue((int64_t)(o.max_size_bytes / 1'000'000));
+    if (n == "mempoolexpiry") return UniValue((int64_t)std::chrono::duration_cast<std::chrono::hours>(o.expiry).count());
+    if (n == "mempoolreplacement") return UniValue(RbfPolicyToString(o.rbf_policy));
+    if (n == "mempooltruc") return UniValue(TrucPolicyToString(o.truc_policy));
+    if (n == "minrelaytxfee") return UniValue(UniValue::VNUM, FormatMoney(o.min_relay_feerate.GetFeePerK()));
+    if (n == "incrementalrelayfee") return UniValue(UniValue::VNUM, FormatMoney(o.incremental_relay_feerate.GetFeePerK()));
+    if (n == "minrelaycoinblocks") return UniValue((int64_t)o.minrelaycoinblocks);
+    if (n == "minrelaymaturity") return UniValue((int64_t)o.minrelaymaturity);
+    if (n == "limitancestorcount") return UniValue((int64_t)o.limits.ancestor_count);
+    if (n == "limitancestorsize") return UniValue((int64_t)(o.limits.ancestor_size_vbytes / 1'000));
+    if (n == "limitdescendantcount") return UniValue((int64_t)o.limits.descendant_count);
+    if (n == "limitdescendantsize") return UniValue((int64_t)(o.limits.descendant_size_vbytes / 1'000));
+    if (n == "rejectparasites") return UniValue(o.reject_parasites);
+    if (n == "rejecttokens") return UniValue(o.reject_tokens);
+    if (n == "subdustfeepenalty") return UniValue(o.subdustfeepenalty);
+    if (n == "acceptnonstdtxn") return UniValue(!o.require_standard);
+    if (n == "acceptunknownwitness") return UniValue(o.acceptunknownwitness);
+    if (n == "permitbaremultisig") return UniValue(o.permit_bare_multisig);
+    if (n == "permitbarepubkey") return UniValue(o.permit_bare_pubkey);
+    if (n == "permitbareanchor") return UniValue(o.permitbareanchor);
+    if (n == "permitbaredatacarrier") return UniValue(o.permitbaredatacarrier);
+    if (n == "permitephemeral") return UniValue(PermitEphemeralToString(o));
+    if (n == "datacarriersize") return UniValue((int64_t)o.max_datacarrier_bytes.value_or(0));
+    if (n == "acceptnonstddatacarrier") return UniValue(o.accept_non_std_datacarrier);
+    if (n == "maxtxlegacysigops") return UniValue((int64_t)o.maxtxlegacysigops);
+    if (n == "dustrelayfee") return UniValue(UniValue::VNUM, FormatMoney(o.dust_relay_feerate_floor.GetFeePerK()));
```
From Code Rabbit:
**Protect the live policy state with a shared synchronization contract**.
These RPC setters write plain globals and pool->m_opts fields without synchronization; the later LOCK(cs_main); LOCK(pool->cs); only covers LimitMempoolSize(), so concurrent readers can race on the same state. Move the in-memory assignments under the existing lock contract, or make the shared globals atomic where lock-free reads are intentional.

kwsantiago replied:
Same as the earlier thread: pre-existing lock-free pattern shared with the GUI (optionsmodel.cpp:1277). A proper fix means atomics across validation/net_processing/GUI, broader than this PR.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,419 @@
+    switch (def.type) {
+    case PolicySettingType::Bool:
+        if (!CoerceBool(value, n.b)) return util::Error{Untranslated(strprintf("%s must be a boolean", def.name))};
+        return n;
+    case PolicySettingType::Int:
+        if (!CoerceInt(value, n.i)) return util::Error{Untranslated(strprintf("%s must be an integer", def.name))};
+        if (n.i < 0) return util::Error{Untranslated(strprintf("%s must not be negative", def.name))};
+        return n;
+    case PolicySettingType::Amount: {
+        const std::string s = value.isNum() ? value.getValStr() : (value.isStr() ? value.get_str() : "");
+        std::optional<CAmount> parsed = ParseMoney(s);
+        if (!parsed) return util::Error{Untranslated(strprintf("%s is not a valid amount", def.name))};
+        n.amount = *parsed;
+        return n;
+    }
+    case PolicySettingType::Number: {
+        const std::string s = value.isNum() ? value.getValStr() : (value.isStr() ? value.get_str() : "");
+        if (!ParseFixedPoint(s, 2, &n.hundredths) || n.hundredths < 0) {
+            return util::Error{Untranslated(strprintf("%s is not a valid number", def.name))};
+        }
+        return n;
```
From Code Rabbit.

Note that this is only a **minor**.  I only mention it for completeness.

**minor**: **Add upper bounds to the int-backed policy settings**.
 Amount and Number already reject negatives; the remaining Int settings still accept any nonnegative int64_t, which can wrap on the unsigned int/int assignments and overflow the * 1000 / * 1'000'000 conversions below.

kwsantiago replied:
Fixed in 1781b8fa: Int settings now carry an upper bound (default UINT32_MAX) so the unsigned and scaled assignments can't wrap.

pdath commented on src/rpc/settings.cpp:
```diff
@@ -0,0 +1,164 @@
+                        {RPCResult::Type::BOOL, "applied", "whether the change was applied"},
+                        {RPCResult::Type::BOOL, "restart_required", "whether a restart is required to take effect"},
+                    }},
+                }},
```
From Code Rabbit.

Note that this is only a **minor**.  I'm not sure I would bother fixing all minors.

**Document the actual failure-result shape**.
Failed entries contain error and omit restart_required, contrary to the declared RPC result schema. This makes generated help inaccurate for clients.

Also applies to: 138-143

kwsantiago replied:
Fixed in 1781b8fa: restart_required and error are now declared optional in the result.

pdath commented on src/rpc/settings.cpp:
```diff
@@ -0,0 +1,164 @@
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid settings: %s", util::Join(errors, "; ")));
+            }
+
+            // Pass 2: apply all (validated) changes.
+            UniValue results(UniValue::VOBJ);
+            for (const std::string& key : keys) {
+                const PolicySettingDef* def = node::FindPolicySetting(key);
+                auto applied = node::UpdatePolicySetting(node, key, settings[key], /*persist=*/true);
+                UniValue entry(UniValue::VOBJ);
+                if (!applied) {
+                    entry.pushKV("applied", false);
+                    entry.pushKV("error", util::ErrorString(applied).original);
+                } else {
+                    entry.pushKV("applied", true);
+                    entry.pushKV("restart_required", def->restart_required);
+                }
+                results.pushKV(key, entry);
+            }
```
From Code Rabbit:

Major: **Make application and persistence atomic across the batch**.
Updates are applied sequentially, so a later UpdatePolicySetting failure leaves earlier settings changed despite the stack’s atomic application/persistence contract. Use a transactional batch operation or rollback all prior updates before returning.

kwsantiago replied:
Fixed in 1781b8fa: mempool availability is checked once before applying, so a mid-batch environmental failure can't leave earlier settings applied. Values are still validated all-or-nothing first.

pdath commented on doc/JSON-RPC-interface.md:
```diff
@@ -202,6 +202,22 @@ or the state of the mempool by an RPC that returned before this RPC. For
+  configuration UI can be generated from the node instead of being hard-coded.
+- `setsettings {"name": value, ...}` validates the whole batch first (nothing is
+  applied if any value is invalid), then applies and persists the valid changes.
+  The result reports whether any change needs a restart to take effect.
```
From Code Rabbit:

Note that this is a **minor**.
**Describe the restart result as per-setting**.
```
-  The result reports whether any change needs a restart to take effect.
+  Each per-setting result reports whether that change needs a restart to take effect.
```

kwsantiago replied:
Fixed in 1781b8fa.

pdath commented on doc/release-notes-k154.md:
```diff
@@ -0,0 +1,22 @@
+RPC and the GUI now change these settings through the same internal interface, so the
+two stay consistent. Setting names match the corresponding `bitcoin.conf` option names.
+
+See `doc/JSON-RPC-interface.md` for the setting list and details.
```
From Code Rabbit.

Note that this is a **minor**.
**Remove the nonexistent setting-list reference**.
The referenced document contains usage details, but directs users to dumpsettings true for the actual setting list.
```
-See `doc/JSON-RPC-interface.md` for the setting list and details.
+See `doc/JSON-RPC-interface.md` for usage details.
```

kwsantiago replied:
Fixed in 1781b8fa.

pdath commented on src/qt/optionsmodel.cpp:
```diff
@@ -823,6 +824,12 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value, const std::
+    auto apply_policy = [&](std::string_view name, const common::SettingsValue& v) {
+        if (changed()) node::UpdatePolicySetting(*node().context(), name, v, /*persist=*/true);
+    };
 
```
From Code Rabbit:

Note that this is a minor.

minor: **Propagate** UpdatePolicySetting() **errors**. apply_policy drops the returned util::Result<void>, so validation/persistence failures are lost and setOption() can still succeed even when the setting wasn’t applied.

kwsantiago replied:
Fixed in 1781b8fa: apply_policy now propagates the error into setOption's return value.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,417 @@
+    auto rwconf = [&](const std::string& key, const std::string& v) { args.ModifyRWConfigFile(key, v); };
+
+    // ---- Globals (no mempool required) ----
+    if (n == "datacarriercost") {
+        g_weight_per_data_byte = (unsigned int)((nv.hundredths * WITNESS_SCALE_FACTOR + 99) / 100);
+        return json("datacarriercost", UniValue(double(nv.hundredths) / 100.0));
+    }
```
From Code Rabbit:

Critical: **Bound** datacarriercost **before arithmetic and narrowing**.

hundredths has no upper bound. Multiplication can overflow int64_t, and otherwise the cast to unsigned int can wrap, producing an unrelated live policy value.

Validate that the rounded result fits unsigned int before assigning it.

pdath commented on src/node/policy_settings.cpp:
```diff
@@ -0,0 +1,417 @@
+        ApplyPermitEphemeralOption(SettingsValue(nv.s), o);
+        return json("permitephemeral", UniValue(nv.s));
+    }
+    if (n == "dustdynamic") {
+        auto parsed = ParseDustDynamicOpt(nv.s, /*max_fee_estimate_blocks=*/1008);
+        if (!parsed) return util::Error{util::ErrorString(parsed)};
+        o.dust_relay_target = parsed->first;
+        o.dust_relay_multiplier = parsed->second;
+        return json("dustdynamic", UniValue(nv.s));
```
From Code Rabbit:

Major: **Recompute the effective dust fee after changing dynamic parameters**.

This updates the target and multiplier but leaves dust_relay_feerate stale until another event recalculates it, so the requested live update is not immediately effective.

```
         o.dust_relay_target = parsed->first;
         o.dust_relay_multiplier = parsed->second;
+        pool->UpdateDynamicDustFeerate();
         return json("dustdynamic", UniValue(nv.s));
```

pdath commented on src/rpc/settings.cpp:
```diff
@@ -0,0 +1,169 @@
+                meta.pushKV("value", *value);
+                meta.pushKV("type", SettingTypeName(def.type));
+                if (def.restart_required) meta.pushKV("restart_required", true);
+                if (auto help = args.GetArgHelpText("-" + def.name)) meta.pushKV("help", *help);
```
From Code Rabbit:

Minor: **Return help for value-taking options**.

Line 74 looks up -maxmempool, but GetArgHelpText() performs an exact lookup and registered options commonly use keys such as -maxmempool=<n> and -minrelaytxfee=<amt>. Detailed output therefore omits help for those settings despite the documented metadata contract. Store the registered argument spelling in the policy definition or canonicalize this lookup; add detailed-output assertions for an integer and an amount setting. ([github.com](vscode-file://vscode-app/c:/Users/pidat/AppData/Local/Programs/Microsoft%20VS%20Code/8a7abeba6e/resources/app/out/vs/code/electron-browser/workbench/workbench.html))

pdath commented on src/node/policy_settings.h:
```diff
@@ -0,0 +1,49 @@
+//! side effects. Used to make setsettings validate the whole batch before applying.
+util::Result<void> ValidatePolicySetting(std::string_view name, const UniValue& value);
+
+//! Validate and apply a setting change: mutate live policy state and write it to
+//! the store the GUI uses for that setting (settings.json or the read-write
+//! config file). Returns an error if the value is invalid or persistence failed.
+util::Result<void> UpdatePolicySetting(NodeContext& node, std::string_view name, const UniValue& value);
```
From Code Rabbit:

Major: **Honor the documented persistence guarantee**.

UpdatePolicySetting mutates live state before PersistJson, so a failed write returns an error after the change is already active. Additionally, the ModifyRWConfigFile wrapper in policy_settings.cpp cannot report failures and always returns success. This can lose settings after restart despite a successful RPC result.

Use an error-reporting persistence API and publish the live mutation only after persistence succeeds, with rollback where necessary.

user475820 commented:
tACK 1781b8fa44412b458141ee8a3ef6052c4b2cde78

Tested:
- macOS Sequoia 15.0 (Apple Silicon), ran rpc_settings.py
- Raspberry Pi (Linux, aarch64), ran rpc_settings.py

Both passed.
